### PR TITLE
mgr/dashboard: Use secure cookies to store JWT Token

### PR DESCRIPTION
**Pull request by aaSharma14**
This PR intends to store the jwt token in secure cookies instead of local storage

Fixes:https://tracker.ceph.com/issues/44591

Signed-off-by: Aashish Sharma <aasharma@redhat.com>
Signed-off-by: Avan Thakkar <athakkar@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/auth-storage.service.spec.ts [hunk 1]
```diff
@@ -34,7 +34,7 @@ describe('AuthStorageService', () => {
   });
 
   it('should be SSO', () => {
-    service.set(username, '', {}, true);
+    service.set(username, {}, true);
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/auth-storage.service.spec.ts:37]: Shouldn't the similar change applied here too ?
 it('should not be SSO', () => {
  ----->  service.set(username, '');	    service.set(username);

**Review by avanthakkar — DISMISSED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/auth-storage.service.spec.ts (hunk 1 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/auth-storage.service.spec.ts:37]: Yes, fixed this. thanks @avanthakkar :+1:

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/controllers/auth.py [hunk 2]
```diff
@@ -38,10 +41,19 @@ def create(self, username, password):
...
+                scheme = ''
             logger.debug('Login successful')
             token = JwtManager.gen_token(username)
             token = token.decode('utf-8')
-            cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
+            cherrypy.response.cookie['token'] = token
+            if 'https' in scheme:
+                cherrypy.response.cookie['token']['secure'] = True
+            cherrypy.response.cookie['token']['HttpOnly'] = True
+            cherrypy.response.cookie['token']['path'] = '/'
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/controllers/auth.py:55]: Shouldn't this the `url_prefix`?

File: src/pybind/mgr/dashboard/controllers/auth.py [hunk 3]
```diff
@@ -91,3 +103,10 @@ def check(self, token):
         return {
             'login_url': self._get_login_url(),
         }
+
+    @RESTController.Collection('GET')
+    def token(self):
+        token = JwtManager.get_token_from_header()
+        return {
+            'token': token
+        }
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/controllers/auth.py:112]: Why is this required?

**Review by epuertat — CHANGES_REQUESTED** [review]
Looks good @aaSharma14 ! I just run a grep and still see some appearances of `access_token` accross the code:
```ts
app.module.ts:  return localStorage.getItem('access_token');
```

## Dashboard :heavy_check_mark:

It works perfectly: the session is kept. Tested with user `admin` and `test`.

## OpenAPI :x:

The existing token-based auth is no longer working:
`curl ... -H 'Authentication: Bearer <jwt token>`

However, the cookies work instead and they might be nice (`-c <file>` for storing the cookies, and `-b <file>` for sending them back to the server):
```
curl -k -X POST "https://localhost:11000/api/auth" -H  "accept: application/vnd.ceph.api.v1.0+json" -H  "Content-Type: application/json" -d "{\"username\":\"admin\",\"password\":\"admin\"}" -c /tmp/cookie.txt
...

curl -k https://localhost:11000/api/pool -H  "accept: application/vnd.ceph.api.v1.0+json" -b /tmp/cookie.txt
...
```
Should we keep the token based authentication too? @alfonsomthd @votdev what do you think?

I think it's pretty 'cheap' for us to keep both mechanisms...

## Grafana ❌ 

Tested with plain HTTP (enabling Insecure Content in Chrome), no Grafana dashboard is displayed and I get this error in JS console:

`Refused to display 'http://172.20.0.6:3000/d/z99hzWtmk/ceph-pools-overview?&refresh=2s&kiosk&from=now-1h&to=now' in a frame because it set 'X-Frame-Options' to 'deny'.`

It can be worked around with [this Chrome extension](https://chrome.google.com/webstore/detail/ignore-x-frame-headers/gleekbfjekiniecknbkamfmkohkpodhe/related?hl=es).

@p-se @pcuzner should we set [`allow_embedding`](https://grafana.com/docs/grafana/latest/administration/configuration/#allow_embedding) in Grafana. Any idea on how to do that?

Are we also providing an easy way to configure `protocol` and certificates in our Grafana container?

## Notifications 

This is unrelated to this PR, but it was recently discussed: notifications are shared across users. The following is the output of `test` user after `admin` user logged out. As seen, most notifications would require admin-level:

image

## SSO :x:

Does not work, but it looks it could be related to the REST API versioning. @avanthakkar could you please 

```
{"status": "415 Unsupported Media Type", "detail": "Unable to find version in request header", "request_id": "206841b7-d745-4fdb-ae41-919d3c7fa64b", "traceback": "Traceback (most recent call last):\n  File \"/usr/lib/python3.6/site-packages/cherrypy/_cprequest.py\", line 638, in respond\n    self._do_respond(path_info)\n  File \"/usr/lib/python3.6/site-packages/cherrypy/_cprequest.py\", line 697, in _do_respond\n    response.body = self.handler()\n  File \"/usr/lib/python3.6/site-packages/cherrypy/lib/encoding.py\", line 219, in __call__\n    self.body = self.oldhandler(*args, **kwargs)\n  File \"/usr/lib/python3.6/site-packages/cherrypy/_cptools.py\", line 230, in wrap\n    return self.newhandler(innerfunc, *args, **kwargs)\n  File \"/ceph/src/pybind/mgr/dashboard/services/exception.py\", line 55, in dashboard_exception_handler\n    raise error\n  File \"/ceph/src/pybind/mgr/dashboard/services/exception.py\", line 46, in dashboard_exception_handler\n    return handler(*args, **kwargs)\n  File \"/usr/lib/python3.6/site-packages/cherrypy/_cpdispatch.py\", line 54, in __call__\n    return self.callable(*self.args, **self.kwargs)\n  File \"/ceph/src/pybind/mgr/dashboard/controllers/__init__.py\", line 691, in inner\n    raise cherrypy.HTTPError(415, \"Unable to find version in request header\")\ncherrypy._cperror.HTTPError: (415, 'Unable to find version in request header')\n", "version": "18.4.0"}
```

**Comment by p-se** [comment]
> @p-se @pcuzner should we set [`allow_embedding`](https://grafana.com/docs/grafana/latest/administration/configuration/#allow_embedding) in Grafana. Any idea on how to do that?

@epuertat As far as I can tell `allow_embedding` is already enabled or mentioned in

- ceph-dev-docker
- cephadm template for octopus (grafana.ini.j2) and
- in the ceph dashboard documentation.

However, it doesn't seem to be enabled in 
- rhcs-dashboard/ceph-dev or
- deepsea (nautilus).

When I read the description I wonder how it worked without having this setting enabled. I further investigated the issue and found out that the setting was added relatively late ([2019-05-07](https://github.com/grafana/grafana/blob/master/CHANGELOG.md#620-beta1-2019-05-07)). I hence think the change added in this PR might not be the cause of the Grafana issue you've mentioned (but note that it depends on the setup and Grafana version used to test it). In any case, without enabling the `allow_embedding` option there's no way to embed Grafana the way Ceph Dashboard does it.

**Comment by epuertat** [comment]
> > @p-se @pcuzner should we set [`allow_embedding`](https://grafana.com/docs/grafana/latest/administration/configuration/#allow_embedding) in Grafana. Any idea on how to do that?
> 
> @epuertat As far as I can tell `allow_embedding` is already enabled or mentioned in
> 
> * ceph-dev-docker
> * cephadm template for octopus (grafana.ini.j2) and
> * in the ceph dashboard documentation.
> 
> However, it doesn't seem to be enabled in
> 
> * rhcs-dashboard/ceph-dev or
> * deepsea (nautilus).
> 
> When I read the description I wonder how it worked without having this setting enabled. I further investigated the issue and found out that the setting was added relatively late ([2019-05-07](https://github.com/grafana/grafana/blob/master/CHANGELOG.md#620-beta1-2019-05-07)). I hence think the change added in this PR might not be the cause of the Grafana issue you've mentioned (but note that it depends on the setup and Grafana version used to test it). In any case, without enabling the `allow_embedding` option there's no way to embed Grafana the way Ceph Dashboard does it.

Thank you @p-se ! I simply worked around it with a Chrome extension to ignore X-Frame headers 🙈 Great to know that it's already in every other place!

File: src/pybind/mgr/dashboard/services/auth.py [hunk 4]
```diff
@@ -67,11 +67,12 @@ def decode_token(cls, token):
...
-        if auth_header is not None:
-            scheme, params = auth_header.split(' ', 1)
-            if scheme.lower() == 'bearer':
-                return params
+        try:
+            auth_header = cherrypy.request.cookie['token'].value
+            if auth_header:
+                return auth_header
+        except KeyError:
+            return None
```
[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/dashboard/services/auth.py:75]: Keeping both ways of authentication:

```suggestion
AUTH_COOKIE_NAME = 'token'
AUTH_HEADER = 'authorization'
AUTH_HEADER_JWT = 'Bearer'
...
        try:
            # use cookie
            return cherrypy.request.cookie[AUTH_COOKIE_NAME].value
        except KeyError:
            try:
                # fall-back: use Authorization header
                return cherrypy.request.headers.get(AUTH_HEADER).split(AUTH_HEADER_JWT)[0].strip()
            except IndexError:
                return None
```

File: src/pybind/mgr/dashboard/controllers/auth.py [hunk 5]
```diff
@@ -1,15 +1,26 @@
...
 from __future__ import absolute_import
 
+import http.cookies
 import logging
+import sys
 
+try:
+    from urlparse import urlparse
+except ImportError:
+    from urllib.parse import urlparse
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/controllers/auth.py:11]: This is trying first the Python 2 syntax, which is the deprecated version. If we want to make this py2 backportable, I think it's better to use [`six urllibparse`](https://six.readthedocs.io/#module-six.moves.urllib.parse), but I'm not sure if we should introduce Py2-3 compatible code in `master`.

What do you thinkg @tchaikov ?

File: src/pybind/mgr/dashboard/openapi.yaml [hunk 6]
```diff
@@ -145,6 +145,26 @@ paths:
...
+          description: Operation exception. Please check the response body for details.
+        '401':
+          description: Unauthenticated access. Please login first.
+        '403':
+          description: Unauthorized access. Please check your permissions.
+        '500':
+          description: Unexpected error. Please check the response body for the stack
+            trace.
+      tags:
+      - Auth
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/openapi.yaml:167]: This spec probably needs to be refreshed.

**Review by epuertat — COMMENTED** [review]
Thank you @avanthakkar! I see you are taking over here, right?

As exposed in [this comment to the tracker issue](https://tracker.ceph.com/issues/44591#note-4), I think this fully resolves both XSS and XSRF attacks, so no need of double cookie as long as we use `SameSite=Strict`.

I see there's still some commented-out code pending to be removed right?

File: src/pybind/mgr/dashboard/controllers/auth.py [hunk 7]
```diff
@@ -38,10 +49,22 @@ def create(self, username, password):
...
+            url_scheme = parsed.scheme
             logger.debug('Login successful')
             token = JwtManager.gen_token(username)
             token = token.decode('utf-8')
-            cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
+            # cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
+            cherrypy.response.cookie['token'] = token
+            if url_scheme == 'https':
+                cherrypy.response.cookie['token']['secure'] = True
+            # cherrypy.response.cookie['token']['HttpOnly'] = True need to dicuss
```
[the remark this answers is not in the document]
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/controllers/auth.py:65]: Yes, I was just testing currently with some changes so pushed the code. I will uncomment this line as well as few commented lines are to be removed

File: src/pybind/mgr/dashboard/openapi.yaml (hunk 6 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/openapi.yaml:167]: Yes no need for it, I'm going to remove it too!

File: src/pybind/mgr/dashboard/controllers/auth.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/controllers/auth.py:65]: And yes +1 for "no need of double cookie"

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/auth.service.ts [hunk 8]
```diff
@@ -48,4 +47,8 @@ export class AuthService {
       window.location.replace(resp.redirect_url);
     });
   }
+
+  getToken() {
+    return this.http.get('api/auth/token');
+  }
```
[the remark this answers is not in the document]
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/auth.service.ts]: Nope, removed it!

File: src/pybind/mgr/dashboard/controllers/auth.py (hunk 5 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/controllers/auth.py]: There is no need for using it any more. I have changed the way of knowing if it's 'https' scheme as suggested by you and with that we no more need to use this piece of code!

Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/auth.py]: Perfect, then. Thanks @avanthakkar !

**Comment by avanthakkar** [comment]
**Update:** All services - Dashboard, Grafana, OpenAPI Docs, SSO are working fine! The only thing remains to be fixed is e2e test failures.

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test api

**Review by aaSharma14 — COMMENTED** [review]
https://github.com/ceph/ceph/blob/4d6d1465ea6d593953564c8119048e10475b4e9c/src/pybind/mgr/dashboard/frontend/cypress/support/commands.ts#L16 - Hey @avanthakkar , I think the trace of *access_token* should be removed from here as well.

**Comment by avanthakkar** [comment]
> https://github.com/ceph/ceph/blob/4d6d1465ea6d593953564c8119048e10475b4e9c/src/pybind/mgr/dashboard/frontend/cypress/support/commands.ts#L16
> - Hey @avanthakkar , I think the trace of _access_token_ should be removed from here as well.

Oh yes. Sure, will do

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by aaSharma14** [comment]
jenkins test api

File: qa/tasks/mgr/dashboard/helper.py [hunk 9]
```diff
@@ -277,25 +277,26 @@ def _request(cls, url, method, data=None, params=None, version=DEFAULT_VERSION):
         url = "{}{}".format(cls._base_uri, url)
         log.info("Request %s to %s", method, url)
         headers = {}
+        cookies = {}
         if cls._token:
             headers['Authorization'] = "Bearer {}".format(cls._token)
-
+            cookies['token'] = cls._token
```
Comment (epuertat) [inline qa/tasks/mgr/dashboard/helper.py:283]: Here I'd choose either. Otherwise if we break either cookie or header authentication, we will never notice (as the other one will work).

What about adding a flag to the request to choose either cookie or header, and testing the header alone (as cookies are our new favorite) in a separate test?

**Review by epuertat — APPROVED** [review]
Great job @avanthakkar! Looks almost ready to me. Just a minor comment over there.

File: qa/tasks/mgr/dashboard/helper.py (hunk 9 above)
Comment (avanthakkar) [reply qa/tasks/mgr/dashboard/helper.py:283]: Yes it can be done, but the problem is we need to duplicate all the testing that we currently do using auth header  for cookies too, because we have to test each & every requests if it works by setting the cookie for token. I mean even if we use some helper function which does the same thing like login but with setting the cookies, how we will actually test in API tests. For e.g. in test_host.py qa task, it first calls a `setUpClass()` which basically sets up the cluster requirement including doing login(using auth header). So if we want to implement the same using cookies, don't we need to do `setUpClass` again!?

I think because for the test suites auth header & cookies work as a replacement, either can be used at one time. So if we want to test for both maybe we need to have separate test files with same test suites but using cookies?

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Review by alfonsomthd — APPROVED** [review]

**Review by nizamial09 — APPROVED** [review]

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by aaSharma14** [comment]
jenkins test make check

**Comment by aaSharma14** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test docs

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins retest this please

**Comment by avanthakkar** [comment]
jenkins test docs

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check
